### PR TITLE
MueLu: split schur and approximate inverse calculation

### PR DESCRIPTION
**Pull request by maxfirmbach**
@trilinos/muelu

## Motivation

The schur complement calculation needs an approximate inverse of one matrix block to
do the respective calculation.

This commit includes the following steps:
- split the schur and inverse calculation
- add an inverse approximation factory
- adjust the testing

I tried to keep backwards compatibility, but this is not possible for all cases ... especially
were certain settings are chosen (e.g. fixing, lumping ...).

The split of those routines will help me with future implementations (e.g. sparse inverse
approximation methods) for the schur complement.

## Stakeholder Feedback

AMG for block matrices with schur complement

## Testing

- modified some existing tests of the schur complement factory to work with the new inverse approximation factory
- added two navier-stokes tests with explicit usage of the inverse approximation factory in the xml files

Joined work with @mayrmt

### CODE REVIEW

**Comment by jhux2** [comment]
@trilinos/muelu

File: packages/muelu/src/Misc/MueLu_InverseApproximationFactory_def.hpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+namespace MueLu {
+
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+  RCP<const ParameterList> InverseApproximationFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::GetValidParameterList() const {
+    RCP<ParameterList> validParamList = rcp(new ParameterList());
+
+    validParamList->set<RCP<const FactoryBase> >("A", NoFactory::getRCP(), "Matrix to build the approximate inverse on.\n");
+
+    validParamList->set<std::string>            ("inverse: approximation type",  "diagonal", "Method used to approximate the inverse.");
+    validParamList->set<bool>                   ("inverse: fixing",              false     , "Fix diagonal by replacing small entries with 1.0");
```
[the remark this answers is not in the document]
Comment (maxfirmbach) [reply packages/muelu/src/Misc/MueLu_InverseApproximationFactory_def.hpp:74]: Small means a hard coded value of 10^-4. Fixing is to my knowledge, by looking at the tests, not used a lot so this is ok for me.

File: packages/muelu/src/Misc/MueLu_InverseApproximationFactory_def.hpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    {
+      auto diag = VectorFactory::Build(A->getRangeMap(), true);
+      A->getLocalDiagCopy(*diag);
+      Ainv = (!fixing ? Utilities::GetInverse(diag) : Utilities::GetInverse(diag, 1e-4, one));
+    }
+    else if(method=="lumping")
+    {
+      auto diag = Utilities::GetLumpedMatrixDiagonal(*A);      
+      Ainv = (!fixing ? Utilities::GetInverse(diag) : Utilities::GetInverse(diag, 1e-4, one));
+    }
```
[the remark this answers is not in the document]
Comment (maxfirmbach) [reply packages/muelu/src/Misc/MueLu_InverseApproximationFactory_def.hpp:117]: I kept it inside as diag is only defined locally and I had this casting issues using it without auto ...

**Comment by maxfirmbach** [comment]
I added unit testing for the diagonal and lumping approaches. Right now I'm using 1D and 2D Laplace matrices to check for the correct maps and values, I guess that is sufficient. What do you think @mayrmt ?

**Comment by mayrmt** [comment]
@maxfirmbach 

>  added unit testing for the diagonal and lumping approaches. Right now I'm using 1D and 2D Laplace matrices to check for the correct maps and values, I guess that is sufficient. What do you think @mayrmt ?

I think this is sufficient. Looks good. Thanks!

File: packages/muelu/test/unit_tests/InverseApproximationFactory.cpp [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+      RCP<Vector> Ainv = level.Get<RCP<Vector> >("Ainv", invapproxFact.get());
+      TEST_EQUALITY(Ainv.is_null(), false);
+      TEST_EQUALITY(Ainv->getMap()->getMinGlobalIndex(), comm->getRank() * int(n/comm->getSize()));
+      TEST_EQUALITY(Ainv->getMap()->getMaxGlobalIndex(), comm->getRank() * int(n/comm->getSize()) + int(n/comm->getSize()-1));
+
+      Teuchos::ArrayRCP<const double> AinvData = Ainv->getData(0);
+      bool bCheck = false;
+      for(int i=0; i<int(n/comm->getSize()); i++)
+        if(std::abs(AinvData[i] - 0.66666666667) < 1e-8 || std::abs(AinvData[i] - 0.5) < 1e-8)
```
Comment (cgcgcg) [inline packages/muelu/test/unit_tests/InverseApproximationFactory.cpp:172]: Does this pass for float and complex<double>?

File: packages/muelu/test/unit_tests/InverseApproximationFactory.cpp [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+      RCP<Vector> Ainv = level.Get<RCP<Vector> >("Ainv", invapproxFact.get());
+      TEST_EQUALITY(Ainv.is_null(), false);
+      TEST_EQUALITY(Ainv->getMap()->getMinGlobalIndex(), comm->getRank() * int(n*n/comm->getSize()));
+      TEST_EQUALITY(Ainv->getMap()->getMaxGlobalIndex(), comm->getRank() * int(n*n/comm->getSize()) + int(n*n/comm->getSize()-1));
+
+      Teuchos::ArrayRCP<const double> AinvData = Ainv->getData(0);
+      bool bCheck = false;
+      for(int i=0; i<int(n*n/comm->getSize()); i++)
+        if(std::abs(AinvData[i] - 0.166667) < 1e-6 || std::abs(AinvData[i] - 0.142857) < 1e-6)
```
Comment (cgcgcg) [inline packages/muelu/test/unit_tests/InverseApproximationFactory.cpp:203]: same question

File: packages/muelu/test/unit_tests/InverseApproximationFactory.cpp (hunk 3 above)
Comment (maxfirmbach) [reply packages/muelu/test/unit_tests/InverseApproximationFactory.cpp:172]: That's actually a good point. For `float` I would say yes, but for `complex` I doubt it. I'll have a look into it.

Comment (cgcgcg) [reply packages/muelu/test/unit_tests/InverseApproximationFactory.cpp]: Could you try it with `float`? It's not enabled by the autotester..

Comment (maxfirmbach) [reply packages/muelu/test/unit_tests/InverseApproximationFactory.cpp]: On my local build it passes with `float` set via `-D Trilinos_ENABLE_FLOAT:BOOL=ON`.

**Comment by maxfirmbach** [comment]
I think regarding the pure code this MR should be fine now. In a next and hopefully final step I would add some information to the user guide / tutorial latex files.

**Comment by maxfirmbach** [comment]
I think adding documentation to doxygen makes more sense than adding information in the user guide, as there is not really a fitting place for this.

**Review by cgcgcg — APPROVED** [review]

File: packages/muelu/src/Misc/MueLu_InverseApproximationFactory_decl.hpp [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+  public:
+    //! @name Constructors/Destructors.
+    //@{
+
+    //! Constructor.
+    InverseApproximationFactory() { }
+
+    //! Destructor.
+    virtual ~InverseApproximationFactory() { }
```
Comment (mayrmt) [inline packages/muelu/src/Misc/MueLu_InverseApproximationFactory_decl.hpp:118]:
```suggestion
    //! Constructor.
    InverseApproximationFactory() = default;

    //! Destructor.
    virtual ~InverseApproximationFactory() = default;
```

File: packages/muelu/src/Misc/MueLu_InverseApproximationFactory_def.hpp [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+  void InverseApproximationFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::DeclareInput(Level& currentLevel) const {
+    Input(currentLevel, "A");
+  }
+
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+  void InverseApproximationFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::Build(Level& currentLevel) const {
+    FactoryMonitor m(*this, "Build", currentLevel);
+
+    typedef Teuchos::ScalarTraits<SC> STS;
```
Comment (mayrmt) [inline packages/muelu/src/Misc/MueLu_InverseApproximationFactory_def.hpp:88]:
```suggestion
    using STS = Teuchos::ScalarTraits<SC>;
```

**Review by mayrmt — APPROVED** [review]
@maxfirmbach Just some tiny suggestions...

**Comment by maxfirmbach** [comment]
That's strange ... it seems like the auto tester is picking up the wrong commit or something like that, would a new force push help?

**Comment by cgcgcg** [comment]
The autotester is just flaky. Unless there are actual errors, we'll just put the retest flag on until it passes.

**Pull request merged**